pennylane/devices/qubit/jaxpr_adjoint.py

```diff
@@ -0,0 +1,163 @@
+
+
+def _operator_forward_pass(eqn, env, ket):
+    """Apply an operator during the forward pass of the adjoint jvp."""
+    invals, tangents = tuple(zip(*(_read(env, var) for var in eqn.invars)))
+    op = eqn.primitive.impl(*invals, **eqn.params)
+    env[eqn.outvars[0]] = (op, ad.Zero(AbstractOperator()))
+
+    if any(not isinstance(t, ad.Zero) for t in tangents[1:]):
+        raise NotImplementedError("adjoint jvp only differentiable parameters in the 0 position.")
+
+    if isinstance(eqn.outvars[0], jax.core.DropVar):
+        return apply_operation(op, ket)
+    if any(not isinstance(t, ad.Zero) for t in tangents):
+        # derivatives of op arithmetic. Should be possible later
+        raise NotImplementedError
+
+    return ket
+
+
+def _measurement_forward_pass(eqn, env, ket):
+    """Perform a measurement during the forward pass of the adjoint jvp."""
+    invals, tangents = tuple(zip(*(_read(env, var) for var in eqn.invars)))
+
+    if any(not isinstance(t, ad.Zero) for t in tangents):  # pragma: no cover
+        # currently prevented by "no differentiable operator arithmetic."
+        # but better safe than sorry to keep this error
+        raise NotImplementedError  # pragma: no cover
+
+    if eqn.primitive.name != "expval_obs":
+        raise NotImplementedError("adjoint jvp only supports expectations of observables.")
+
+    mp = eqn.primitive.impl(*invals, **eqn.params)
+    bra = apply_operation(mp.obs, ket)
+    result = jnp.real(jnp.sum(jnp.conj(bra) * ket))
+    env[eqn.outvars[0]] = (result, None)
+    return bra
+
+
+def _other_prim_forward_pass(eqn: jax.core.JaxprEqn, env: dict) -> None:
+    """Handle any equation that is not an operator or measurement eqn.
+
+    Maps outputs back to the environment
+    """
+    invals, tangents = tuple(zip(*(_read(env, var) for var in eqn.invars)))
+    if eqn.primitive not in ad.primitive_jvps:
+        raise NotImplementedError(
+            f"Primitive {eqn.primitive} does not have a jvp rule and is not supported.."
+        )
+    outvals, doutvals = ad.primitive_jvps[eqn.primitive](invals, tangents, **eqn.params)
+    if not eqn.primitive.multiple_results:
+        outvals = [outvals]
+        doutvals = [doutvals]
+    for var, v, dv in zip(eqn.outvars, outvals, doutvals, strict=True):
+        env[var] = (v, dv)
+
+
+def _forward_pass(jaxpr: jax.core.Jaxpr, env: dict, num_wires: int):
+    """Calculate the forward pass of an adjoint jvp calculation."""
+    bras = []
+    ket = create_initial_state(range(num_wires))
+
+    for eqn in jaxpr.eqns:
+        if getattr(eqn.primitive, "prim_type", "") == "operator":
+            ket = _operator_forward_pass(eqn, env, ket)
+
+        elif getattr(eqn.primitive, "prim_type", "") == "measurement":
+            bra = _measurement_forward_pass(eqn, env, ket)
+            bras.append(bra)
+        else:
+            _other_prim_forward_pass(eqn, env)
```

Comment:
```suggestion
    for eqn in jaxpr.eqns:
        prim_type = getattr(eqn.primitive, "prim_type", "")
        if prim_type == "operator":
            ket = _operator_forward_pass(eqn, env, ket)
        elif prim_type == "measurement":
            bras.append(_measurement_forward_pass(eqn, env, ket))
        else:
            _other_prim_forward_pass(eqn, env)
```

What do you think about this version with only one call to `getattr`?

Reply:
I investigated the difference between the two yesterday when I was messing around with performance improvements. There was no change between the two.

Reply:
Yeah, if there is a difference at all it's probably even hard to measure it